pkg/controllers/job/job_controller.go

```diff
@@ -197,27 +217,96 @@ func (cc *Controller) Run(stopCh <-chan struct{}) {
 		cc.svcSynced, cc.cmdSynced, cc.pvcSynced, cc.pcSynced)
 
 	go wait.Until(cc.handleCommands, 0, stopCh)
-	go wait.Until(cc.worker, 0, stopCh)
+	var i uint32
+	for i = 0; i < workers; i++ {
+		go func(num uint32) {
+			wait.Until(
+				func() {
+					cc.worker(num)
+				},
+				time.Second,
+				stopCh)
+		}(i)
+
+	}
 
 	go cc.cache.Run(stopCh)
 
 	glog.Infof("JobController is running ...... ")
 }
 
-func (cc *Controller) worker() {
-	for cc.processNextReq() {
+func (cc *Controller) worker(i uint32) {
+
+	glog.Infof("worker %d start ...... ", i)
+
+	for cc.processNextReq(i) {
+	}
+}
+
+// TODO we may need to make this sharding more proper if required
+func (cc *Controller) belongsToThisRoutine(key string, count uint32) bool {
+	var hashVal hash.Hash32
+	var val uint32
+
+	hashVal = fnv.New32()
+	hashVal.Write([]byte(key))
+
+	val = hashVal.Sum32()
+
+	if val%cc.workers == count {
+		return true
 	}
+
+	return false
+}
+
+// TODO we may need to make this sharding more proper if required
+func (cc *Controller) getWorkerID(key string) uint32 {
+	var hashVal hash.Hash32
+	var val uint32
+
+	hashVal = fnv.New32()
+	hashVal.Write([]byte(key))
+
+	val = hashVal.Sum32()
+
+	return val % cc.workers
 }
 
-func (cc *Controller) processNextReq() bool {
-	obj, shutdown := cc.queue.Get()
+func (cc *Controller) processNextReq(count uint32) bool {
+	obj, shutdown := cc.queueList[count].Get()
 	if shutdown {
 		glog.Errorf("Fail to pop item from queue")
 		return false
 	}
 
 	req := obj.(apis.Request)
-	defer cc.queue.Done(req)
+	defer cc.queueList[count].Done(req)
+
+	key := jobcache.JobKeyByReq(&req)
+	// Later we can remove this code if we want
+	if !cc.belongsToThisRoutine(key, count) {
+		glog.Errorf("should not occur The job does not belongs to this routine key:%s, worker:%d...... ", key, count)
+		cc.queueList[count].AddRateLimited(req)
+		return true
+	}
+
+	// prevent multi threads processing the same job simultaneously.
+	cc.jobsLock.Lock()
```

Comment:
This would not happen if you do sharding in the source.

Reply:
agreed, will  remove the code